src/identity/manager.rs

```diff
@@ -32,16 +37,141 @@ impl fmt::Display for Identity {
+                                    tx_option = Some(tx);
+                                }
+                            };
+                        }
+                        Some(cert_rcvr) => {
+                            info!("Got cached cert receiver.");
+                            cache_rx = Some(cert_rcvr.clone());
+                        }
+                    }
+                }
+
+                // We made a transmitter, so fetch the cert and send the result.
+                if tx_option.is_some() {
+                    let certs = self.client.fetch_certificate(id.clone()).await;
+                    match certs {
+                        Ok(c) => {
+                            let tx = tx_option.unwrap();
+                            match tx.send(Some(c.clone())) {
+                                Ok(_) => {
+                                    tokio::spawn(SecretManager::refresh_handler(
+                                        id.clone(),
+                                        self.clone(),
+                                        c.get_duration_until_refresh(),
+                                        tx,
+                                    ));
+                                }
+                                Err(e) => {
+                                    warn!("Failed to send fetched certs on channel: {:?}", e);
+                                    let mut locked_cache = self.cache.write().unwrap();
+                                    locked_cache.remove(id);
+                                }
+                            }
+
+                            return Ok(c);
+                        }
+                        Err(e) => {
+                            let mut locked_cache = self.cache.write().unwrap();
+                            locked_cache.remove(id);
+                            return Err(e);
+                        }
+                    }
+                }
+
+                // We found a channel receiver in the cache, so use it to get the cert.
+                loop {
+                    let current_cert_info = cache_rx.as_mut().unwrap();
+                    {
+                        let current_cert = current_cert_info.borrow_and_update();
+                        if current_cert.is_some() {
+                            return Ok(current_cert.clone().unwrap());
+                        }
+                    }
+                    // "None" means CA request in progress.  Wait for it.
+                    match current_cert_info.changed().await {
+                        Err(_) => {
+                            // CA request failed or other error.
+                            warn!("Cert sender has been dropped.");
+                            break;
+                        }
+                        Ok(_) => {
+                            continue;
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
Can we add some unit tests to this? We should be able to mock out the actual CA call with a fake CaClient (might require some refactoring to allow CaClient to be generic)